src/client/mod.rs

```diff
@@ -157,78 +158,55 @@ impl APIClient {
-                            Ok(Some(chunk)) => {
-                                // append it to our current buffer
-                                buff.extend_from_slice(&chunk);
-
-                                // If it contains a new line, we have a full item somewhere in our buffer
-                                // Note: It seems that the kube api escapes all new line characters excepting the ones that
-                                // separate distinct events.
-                                if chunk.contains(&b'\n') {
-                                    // Create a new buffer for our next iteration
-                                    // to move any partial lines to
-                                    // We an safely reuse the items vec as we know it's empty here.
-                                    let mut new_buff = Vec::new();
-
-                                    for line in buff.split(|x| x == &b'\n') {
-                                        match serde_json::from_slice(&line) {
-                                            Ok(val) => items.push(Ok(val)),
-                                            Err(e) if e.is_eof() => {
-                                                // If this is an EOF error then we do not have a complete
-                                                // kube object, so push the current line into our new buffer
-                                                // and add the newline character we stripped as part of the iterator.
-                                                new_buff.extend_from_slice(&line);
-                                                new_buff.push(b'\n');
-                                            }
-                                            Err(e) => items.push(Err(Error::SerdeError(e))),
+        let stream = futures::stream::unfold((res, Vec::new()), |(mut resp, mut buff)| {
+            async {
+                loop {
+                    match resp.chunk().await {
+                        Ok(Some(chunk)) => {
+                            buff.extend_from_slice(&chunk);
+
+                            // If we've encountered a newline, see if we have any items to yield
+                            if chunk.contains(&b'\n') {
+                                let mut new_buff = Vec::new();
+                                let mut items = Vec::new();
+
+                                // Split on newlines
+                                for line in buff.split(|x| x == &b'\n') {
+                                    match serde_json::from_slice(&line) {
+                                        Ok(val) => items.push(Ok(val)),
+                                        Err(e) if e.is_eof() => {
+                                            // In this case we've received only a partial object
+                                            // so store it for use in the next iteration
+                                            new_buff.extend_from_slice(&line);
```

Comment:
We might get more than one chunk here which still isn't a complete line in an extreme case. If this were to happen, we would not parse it until the next event (which could take ages).

It's probably better to parse in a `while more_remaining_data; try to parse current_buf + next_line_chunk`.

Reply:
A way to test you're getting all events (and emulate a chattier apiserver), slap a sleep in  the middle of the unfold:

```diff
diff --git a/src/client/mod.rs b/src/client/mod.rs
index 4226bfeb..64dd5c62 100644
--- a/src/client/mod.rs
+++ b/src/client/mod.rs
@@ -167,6 +167,9 @@ impl APIClient {
                     match resp.chunk().await {
                         Ok(Some(chunk)) => {
                             buff.extend_from_slice(&chunk);
+                            use futures_timer::Delay;
+                            use std::time::Duration;
+
 
                             // If we've encountered a newline, see if we have any items to yield
                             if chunk.contains(&b'\n') {
@@ -175,12 +178,16 @@ impl APIClient {
 
                                 // Split on newlines
                                 for line in buff.split(|x| x == &b'\n') {
+                                    let dur = Duration::from_secs(10);
+                                    Delay::new(dur).await;
+
                                     match serde_json::from_slice(&line) {
                                         Ok(val) => items.push(Ok(val)),
                                         Err(e) if e.is_eof() => {
                                             // In this case we've received only a partial object
                                             // so store it for use in the next iteration
                                             new_buff.extend_from_slice(&line);
+
                                         }
                                         Err(e) => {
                                             warn!(
```

Reply:
Actually, this is exposing really weird behaviour from the http interface. Even if you sleep just before the `chunk await`, you still end up dropping a huge amount of events with this. Not sure why. Maybe they are being dropped at like the TCP layer, or maybe the code here is wrong (although it seems correct to me now).